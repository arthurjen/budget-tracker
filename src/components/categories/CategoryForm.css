:local(.categoryForm) {

  display: flex;
  justify-content: center;
  align-items: center;

  height: 80px;
  
  

  .inputs {
    display: flex;
    flex-direction: column;
  }

  input {
    border: 1px solid rgb(226, 145, 23);
    height: 25px;
    width: 150px;
    font-size: 1.2em;
    margin: 1px;
  }
  

  .buttons {
    display: flex;
    justify-content: space-around;
    align-items: center;
  }

  button {
    border: 1px solid rgb(226, 145, 23);
    font-size: 20px;
    width: 60px;
    height: 60px;
    background: white;
  }

  .fa-ban, .fa-trash-alt {
    color: red;
  }

  .fa-plus, .fa-check {
    color: green
  }
}