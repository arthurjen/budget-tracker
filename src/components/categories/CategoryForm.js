import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styles from './CategoryForm.css';

class CategoryForm extends PureComponent {
  
  state = {
    name: '',
    budget: '',
    timestamp: null,
    id: null
  };

  static propTypes = {
    category: PropTypes.object,
    onComplete: PropTypes.func.isRequired,
    onCancel: PropTypes.func,
    onRemove: PropTypes.func
  };

  handleSubmit = event => {
    event.preventDefault();
    const { name, budget, id, timestamp } = this.state;
    if(!name || !budget) return;
    let category = { name, budget };
    if(id) category = { ...category, id, timestamp };
    this.props.onComplete(category);
  };

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  componentDidMount() {
    const { category } = this.props;
    if(!category) return;
    this.setState(category);
  }

  onDelete = event => {
    event.preventDefault();
    const { onRemove } = this.props;
    const { id } = this.state;
    onRemove(id);
  };


  render() { 
    const { name, budget, id } = this.state;
    const { onCancel } = this.props;

    return (
      <form className={styles.categoryForm} onSubmit={this.handleSubmit}>
        
        <section className="inputs">
          <label>
            <input name="name" placeholder="Category" value={name} onChange={this.handleChange}/>
          </label>
          <label>
            <input name="budget" type="number" placeholder="Maximum $$$" value={budget} onChange={this.handleChange}/>
          </label>
        </section>
        {id &&
          <button type="remove" onClick={this.onDelete}><i className="far fa-trash-alt" ></i></button>
        }
        <button type="button" onClick={onCancel}><i className="fas fa-ban"></i></button>
        <button type="submit">
          {id 
            ? <i className="fas fa-check"></i>
            : <i className="fas fa-plus"></i>
          }
        </button>
      </form>
    );
  }
}


 
export default CategoryForm;