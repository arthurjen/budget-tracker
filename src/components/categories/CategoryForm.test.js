import CategoryForm from './CategoryForm';
import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

describe('Category Form', () => {

  it('renders add if no category prop', () => {
    const add = jest.fn();
    const promise = Promise.resolve();
    add.mockReturnValueOnce(promise);

    const wrapper = mount(<CategoryForm onComplete={add}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();

    const category = {
      name: 'Test Category',
      budget: 666,
    };

    wrapper.find('input[name="name"]').simulate('change', {
      target: {
        name: 'name',
        value: category.name
      }
    });

    wrapper.find('input[name="budget"]').simulate('change', {
      target: {
        name: 'budget',
        value: category.budget
      }
    });

    wrapper.find('button[type="submit"]').simulate('submit');

    const calls = add.mock.calls;
    expect(calls.length).toBe(1);
    expect(calls[0][0]).toEqual(category);

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('renders edit if passed a category prop', () => {
    const update = jest.fn();
    const promise = Promise.resolve();
    update.mockReturnValueOnce(promise);
    const toggleEdit = jest.fn();

    const category = {
      id: '123456',
      name: 'Test Category',
      budget: 666,
      timestamp: 'test date'
    };

    const wrapper = mount(<CategoryForm
      onComplete={update}
      onCancel={toggleEdit}
      category={category}
    />);

    wrapper.find('input[name="budget"]').simulate('change', {
      target: {
        name: 'budget',
        value: 999
      }
    });

    wrapper.find('button[type="submit"]').simulate('submit');

    const calls = update.mock.calls;
    expect(calls.length).toBe(1);
    expect(calls[0][0]).toEqual({
      ...category,
      budget: 999
    });

    
  });
});