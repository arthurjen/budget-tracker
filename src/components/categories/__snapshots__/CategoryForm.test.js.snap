// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category Form renders add if no category prop 1`] = `
<CategoryForm
  onComplete={[MockFunction]}
>
  <form
    className="categoryForm"
    onSubmit={[Function]}
  >
    <section
      className="inputs"
    >
      <label>
        <input
          name="name"
          onChange={[Function]}
          placeholder="Category"
          value=""
        />
      </label>
      <label>
        <input
          name="budget"
          onChange={[Function]}
          placeholder="Maximum $$$"
          type="number"
          value=""
        />
      </label>
    </section>
    <button
      type="button"
    >
      <i
        className="fas fa-ban"
      />
    </button>
    <button
      type="submit"
    >
      <i
        className="fas fa-plus"
      />
    </button>
  </form>
</CategoryForm>
`;

exports[`Category Form renders add if no category prop 2`] = `
<CategoryForm
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "budget": 666,
            "name": "Test Category",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <form
    className="categoryForm"
    onSubmit={[Function]}
  >
    <section
      className="inputs"
    >
      <label>
        <input
          name="name"
          onChange={[Function]}
          placeholder="Category"
          value="Test Category"
        />
      </label>
      <label>
        <input
          name="budget"
          onChange={[Function]}
          placeholder="Maximum $$$"
          type="number"
          value={666}
        />
      </label>
    </section>
    <button
      type="button"
    >
      <i
        className="fas fa-ban"
      />
    </button>
    <button
      type="submit"
    >
      <i
        className="fas fa-plus"
      />
    </button>
  </form>
</CategoryForm>
`;
